prov/gni/include/gnix_buddy_allocator.h

```diff
@@ -0,0 +1,177 @@
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _GNIX_BUDDY_ALLOCATOR_H_
+#define _GNIX_BUDDY_ALLOCATOR_H_
+
+#include "fi_list.h"
+#include "gnix_bitmap.h"
+#include "gnix_util.h"
+#include <math.h>
+#include <stdlib.h>
+
+#define MIN_BLOCK_SIZE 16
+
+/* evaluates to zero if X is not a power of two, otherwise evaluates to X - 1 */
+#define IS_NOT_POW_TWO(X) (((X) & (~(X) + 1)) ^ (X))
+
+/* Find the block size (in bytes) required for allocating LEN bytes */
+#define BLOCK_SIZE(LEN, MIN) ((LEN) <= (MIN) ? (MIN) :\
+			      (IS_NOT_POW_TWO(LEN)) ? (((LEN) << 1) & ~(LEN)) :\
+			      (LEN))
+
+/* Find the bitmap index for block X */
+#define BITMAP_INDEX(X, BASE, MIN, LEN) (size_t) ((size_t) ((X) - (BASE)) /\
+					 (MIN) + 2 * log2((LEN) / (MIN)))
+
+/*
+ * The following macro doesn't work when the base address starts at zero.
+ * #define BUDDY(X, SIZE_X, BASE) (void *) ((size_t) (X) ^ (SIZE_X))
+ */
+
+/* Find the address of X's buddy block:
+ * If the "index" of block X is even then the buddy must be to the right of X,
+ * otherwise the buddy is to the left of X.
+ */
+#define BUDDY(X, LEN, BASE) (void *) ((((size_t) (BASE) - (size_t) (X)) /\
+				       (LEN)) % 2 ? (size_t) (X) - (LEN) :\
+				      (size_t) (X) + (LEN))
+
+/* Calculate the offset of a free block, OFFSET = MIN * 2^MULT. */
+#define OFFSET(MIN, MULT) ((MIN) * (1 << (MULT)))
+
+/* Find the index into the free list with block size LEN. */
+#define LIST_INDEX(LEN, MIN) (size_t) (log2((LEN) / (double) (MIN)))
+
+/**
+ * Structure representing a buddy allocator.
+ *
+ * @var base		The base address of the buffer being managed.
+ * @var len		The length of the buffer the buddy allocator is managing.
+ * @var max		The largest chunk of memory that can be allocated.
+ *
+ * @var nlists		The number of free lists.
+ * @var lists		The array of free lists ordered from smallest block size.
+ * at index 0 to largest block size at index nlists - 1.
+ *
+ * @var bitmap		Each bit is 1 if the block is allocated or split,
+ * otherwise the bit is 0.
+ *
+ * @var lock		The buddy alloc handle lock.
+ */
+typedef struct gnix_buddy_alloc_handle {
+	void *base;
+	size_t len;
+	size_t max;
+
+	size_t nlists;
+	struct dlist_entry *lists;
+
+	gnix_bitmap_t bitmap;
+
+	fastlock_t lock;
+} handle_t;
```

Comment:
Ok, I agree.

> On Jan 7, 2016, at 5:11 PM, Zach Tiffany <notifications@github.com> wrote:
> 
> In prov/gni/include/gnix_buddy_allocator.h:
> 
> > + * otherwise the bit is 0.
> > + *
> > + * @var lock		The buddy alloc handle lock.
> > + */
> > +typedef struct gnix_buddy_alloc_handle {
> > +	void *base;
> > +	size_t len;
> > +	size_t max;
> > +
> > +	size_t nlists;
> > +	struct dlist_entry *lists;
> > +
> > +	gnix_bitmap_t bitmap;
> > +
> > +	fastlock_t lock;
> > +} handle_t;
> I think handle_t is too generic a type name.
> 
> —
> Reply to this email directly or view it on GitHub.
>